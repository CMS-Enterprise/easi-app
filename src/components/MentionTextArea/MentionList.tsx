/* MentionList renders the TipTap suggestion dropdown in addition to defining
defining keyboard events */

import React, {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useState
} from 'react';
import { useTranslation } from 'react-i18next';
import { TagType } from 'gql/generated/graphql';

import Spinner from 'components/Spinner';
import {
  MentionListOnKeyDown,
  MentionSuggestionProps
} from 'types/discussions';

import './index.scss';

export const SuggestionLoading = () => {
  return (
    <div className="items width-full padding-1">
      <Spinner size="small" />
    </div>
  );
};

/** Handler dropdown scroll event on keypress */
const scrollIntoView = () => {
  const selectedElm = document.querySelector('.is-selected');
  selectedElm?.scrollIntoView({ block: 'nearest' });
};

/** Renders the list of suggestions within `MentionTextArea` */
const MentionList = forwardRef<MentionListOnKeyDown, MentionSuggestionProps>(
  (props, ref) => {
    const { t } = useTranslation('general');

    const [selectedIndex, setSelectedIndex] = useState<number>(0);

    /** Sets the selected mention within the editor props */
    const selectItem = (index: number) => {
      const item = props.items[index];

      if (item) {
        props.command({
          'tag-type': item.tagType,
          label: item.displayName,
          'data-label': item.displayName,
          'data-id-db': item.tagType === TagType.USER_ACCOUNT ? item.id : ''
        });
      }
    };

    const upHandler = () => {
      setSelectedIndex(
        (selectedIndex + props.items?.length - 1) % props.items?.length
      );
      scrollIntoView();
    };

    const downHandler = () => {
      setSelectedIndex((selectedIndex + 1) % props.items?.length);
      scrollIntoView();
    };

    const enterHandler = () => {
      selectItem(selectedIndex);
    };

    useEffect(() => setSelectedIndex(0), [props.items]);

    useImperativeHandle(ref, () => ({
      onKeyDown: ({ event }) => {
        if (
          event.key === 'ArrowUp' ||
          (event.shiftKey && event.key === 'Tab')
        ) {
          upHandler();
          return true;
        }

        if (
          event.key === 'ArrowDown' ||
          (!event.shiftKey && event.key === 'Tab')
        ) {
          downHandler();
          return true;
        }

        if (event.key === 'Enter') {
          enterHandler();
          return true;
        }

        return false;
      }
    }));

    return (
      <div className="items">
        {props.items?.length ? (
          props.items?.map((item, index) => (
            <button
              className={`item ${index === selectedIndex ? 'is-selected' : ''}`}
              key={item.displayName}
              id={item.displayName}
              type="button"
              onClick={() => selectItem(index)}
            >
              {item.displayName}
            </button>
          ))
        ) : (
          <span className="item padding-x-1">{t('noResults')}</span>
        )}
      </div>
    );
  }
);

export default MentionList;
