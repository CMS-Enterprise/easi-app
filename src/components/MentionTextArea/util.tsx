// Possible Util to extract only mentions from content
// eslint-disable-next-line import/prefer-default-export
export const getMentions = (data: any) => {
  const mentions: any = [];

  data?.content?.forEach((para: any) => {
    para?.content?.forEach((content: any) => {
      if (content?.type === 'mention') {
        mentions.push(content?.attrs);
      }
    });
  });

  return mentions;
};
