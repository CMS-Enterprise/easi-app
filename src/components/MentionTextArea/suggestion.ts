import { ReactRenderer } from '@tiptap/react';
import { SuggestionOptions } from '@tiptap/suggestion';
import tippy, { GetReferenceClientRect, Instance } from 'tippy.js';

import {
  MentionAttributes,
  MentionListOnKeyDown,
  MentionSuggestion,
  MentionSuggestionProps
} from 'types/discussions';

import MentionList, { SuggestionLoading } from './MentionList';

/* Returns the current textarea/RTE editor dimension to append the Mentionslist dropdown
MentionList should have the same width as this parent clientRect */
const getClientRect = ({
  editor,
  clientRect
}: MentionSuggestionProps): GetReferenceClientRect => {
  const { element } = editor.options;
  const rect = element.getBoundingClientRect();
  const mentionRect = clientRect?.();

  return () =>
    new DOMRect(
      rect?.left,
      mentionRect?.y,
      mentionRect?.width,
      mentionRect?.height
    );
};

const suggestion: Omit<
  SuggestionOptions<MentionSuggestion, MentionAttributes>,
  'editor'
> = {
  allowSpaces: true,
  render: () => {
    let reactRenderer: ReactRenderer<
      MentionListOnKeyDown,
      MentionSuggestionProps
    >;

    let spinner: Partial<Instance>;
    let popup: Partial<Instance>;

    return {
      // If we had async initial data - load a spinning symbol until onStart gets called
      // We have hardcoded in memory data for current implementation, doesn't currently get called
      onBeforeStart: props => {
        if (!props.clientRect) {
          return;
        }

        reactRenderer = new ReactRenderer(SuggestionLoading, {
          props,
          editor: props.editor
        });

        [spinner] = tippy('body', {
          getReferenceClientRect: getClientRect(props),
          appendTo: props.editor.options.element,
          content: reactRenderer.element,
          showOnCreate: true,
          interactive: false,
          trigger: 'manual',
          placement: 'bottom-start'
        });
      },

      // Render any available suggestions when mention trigger is first called - @
      onStart: props => {
        if (!props.clientRect) {
          return;
        }

        spinner.hide?.();

        reactRenderer = new ReactRenderer(MentionList, {
          props,
          editor: props.editor
        });

        [popup] = tippy('body', {
          getReferenceClientRect: getClientRect(props),
          appendTo: props.editor.options.element,
          content: reactRenderer.element,
          showOnCreate: true,
          interactive: true,
          trigger: 'manual',
          placement: 'bottom-start'
        });
      },

      // When async data/suggestions return, hide the spinner and show the updated list
      onUpdate: props => {
        reactRenderer.updateProps(props);

        if (!props.clientRect) {
          return;
        }

        popup.setProps?.({
          getReferenceClientRect: getClientRect(props)
        });

        spinner.setProps?.({
          getReferenceClientRect: getClientRect(props)
        });

        spinner.hide?.();

        popup.show?.();
      },

      // If a valid character key, render the spinner until onUpdate gets called to rerender updated list
      onKeyDown: props => {
        if (props.event.key === 'Escape') {
          popup.hide?.();
          spinner.hide?.();

          return true;
        }

        if (props.event.key.length === 1 || props.event.key === 'Backspace') {
          popup.hide?.();
          spinner.show?.();
        }

        return !!reactRenderer?.ref && reactRenderer.ref.onKeyDown(props);
      },

      onExit() {
        popup.destroy?.();
        spinner.destroy?.();
        reactRenderer.destroy();
      }
    };
  }
};

export default suggestion;
