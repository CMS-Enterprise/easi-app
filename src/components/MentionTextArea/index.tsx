import React, { useEffect, useMemo, useState } from 'react';
import { useTranslation } from 'react-i18next';
import Document from '@tiptap/extension-document';
import Mention from '@tiptap/extension-mention';
import Paragraph from '@tiptap/extension-paragraph';
import Text from '@tiptap/extension-text';
import {
  EditorContent,
  NodeViewWrapper,
  ReactNodeViewRenderer,
  useEditor
} from '@tiptap/react';
import { Icon } from '@trussworks/react-uswds';
import classNames from 'classnames';

import Alert from 'components/shared/Alert';
import IconButton from 'components/shared/IconButton';

import suggestion from './suggestion';
import { getMentions } from './util';

import './index.scss';

/* The rendered Mention after selected from MentionList
This component can be any react jsx component, but must be wrapped in <NodeViewWrapper />
Attrs of selected mention are accessed through node prop */
const MentionComponent = ({ node }: { node: any }) => {
  const { label } = node.attrs;

  // Label may return null if the text was truncated by <TruncatedText />
  // In this case don't render the mention, and shift the line up by the height of the non-rendered label
  if (!label) {
    return <div className="margin-top-neg-4" />;
  }

  return (
    <NodeViewWrapper className="react-component display-inline">
      <span className="tiptap mention">{`@${label}`}</span>
    </NodeViewWrapper>
  );
};

/* Extended TipTap Mention class with additional attributes
Additionally sets a addNodeView to render custo JSX as mention */
const CustomMention = Mention.extend({
  atom: true,
  selectable: true,
  addAttributes() {
    return {
      ...this.parent?.(),
      'data-id-db': {
        default: ''
      },
      'tag-type': {
        default: ''
      }
    };
  },
  addNodeView() {
    return ReactNodeViewRenderer(MentionComponent);
  }
});

type MentionTextAreaProps = {
  id: string;
  setFieldValue?: (value: string) => void;
  editable?: boolean;
  disabled?: boolean;
  initialContent?: string;
  /** Truncate text with read more/less button for non-editable text */
  truncateText?: boolean;
  className?: string;
};

/**
 * Rich text area component with functionality to tag users or teams
 */
const MentionTextArea = React.forwardRef<HTMLDivElement, MentionTextAreaProps>(
  (
    {
      id,
      setFieldValue,
      editable = false,
      disabled,
      initialContent,
      truncateText,
      className
    },
    ref
  ) => {
    const { t } = useTranslation('discussions');

    const [textExpanded, setTextExpanded] = useState(false);

    /** Alert shown below field when tagging user or user group */
    const [tagAlert, setTagAlert] = useState<boolean>(false);

    /** Mock users array for testing until tagging functionality is implemented  */
    const fetchUsers = ({ query }: { query: string }) => {
      return [
        { username: 'a', displayName: 'Admin lead', tagType: 'other' },
        {
          username: 'b',
          displayName: 'Governance Admin Team',
          tagType: 'other'
        },
        {
          username: 'c',
          displayName: 'Governance Review Board (GRB)',
          tagType: 'other'
        },
        {
          username: 'OSYC',
          displayName: 'Grant Eliezer',
          tagType: 'user'
        },
        {
          username: 'MKCK',
          displayName: 'Forest Brown',
          tagType: 'user'
        },
        {
          username: 'PJEA',
          displayName: 'Janae Stokes',
          tagType: 'user'
        }
      ];
    };

    /** Character limit when truncating text in non-editable text area */
    const truncatedTextCharLimit = 275;

    const textIsTruncated = useMemo(() => {
      if (editable || !initialContent) return false;

      return truncateText && initialContent.length > truncatedTextCharLimit;
    }, [initialContent, truncateText, truncatedTextCharLimit, editable]);

    /** Editor content */
    const content = useMemo(() => {
      // If no initial content, set to empty string
      if (!initialContent) return '';

      // If text is not truncated or truncated and expanded, return `initialContent`
      if (!textIsTruncated || textExpanded) return initialContent;

      // Return truncated text with ellipses
      return `${initialContent.slice(0, truncatedTextCharLimit)} ...`;
    }, [textIsTruncated, truncatedTextCharLimit, textExpanded, initialContent]);

    /** Tiptap editor instance */
    const editor = useEditor(
      {
        editable: editable && !disabled,
        editorProps: {
          attributes: {
            id,
            role: 'textbox',
            'aria-label': t('Rich text area')
          }
        },
        extensions: [
          Document,
          Paragraph,
          Text,
          CustomMention.configure({
            HTMLAttributes: {
              class: 'mention'
            },
            suggestion: {
              ...suggestion,
              items: fetchUsers
            }
          })
        ],
        onUpdate: ({ editor: input }) => {
          const inputContent = input?.getHTML();
          const inputText = input?.getText();

          if (setFieldValue) {
            // Prevents editor from setting value to '<p></p>' when user deletes all text
            if (inputText === '') {
              setFieldValue('');
              return;
            }

            setFieldValue(inputContent);
          }
        },
        // Sets an alert if a mention is selected, and users/teams will be emailed
        onSelectionUpdate: ({ editor: input }: any) => {
          setTagAlert(!!getMentions(input?.getJSON()).length);
        },
        content
      },
      [textExpanded]
    );

    /** Clear editor content when field is reset */
    useEffect(() => {
      if (editable) {
        if (
          !initialContent ||
          // Check if value is empty string and editor is not already reset
          (initialContent.length === 0 && initialContent !== editor?.getText())
        ) {
          editor?.commands.clearContent();
        }
      }
    }, [editor, initialContent, editable]);

    return (
      <>
        <EditorContent
          ref={ref}
          tabIndex={editable ? -1 : undefined}
          editor={editor}
          className={classNames(className, {
            'tiptap__editable usa-textarea': editable,
            'tiptap__readonly font-body-sm line-height-body-5': !editable,
            'tiptap__readonly--truncated': truncateText && !textExpanded
          })}
        />

        {
          // Read more/less button for truncated text
          textIsTruncated && (
            <IconButton
              icon={textExpanded ? <Icon.ExpandLess /> : <Icon.ExpandMore />}
              type="button"
              onClick={() => setTextExpanded(!textExpanded)}
              iconPosition="after"
              className="margin-bottom-205"
              unstyled
            >
              {textExpanded ? t('general:readLess') : t('general:readMore')}
            </IconButton>
          )
        }

        {tagAlert && editable && (
          <Alert type="info" slim>
            {t('general.alerts.saveDiscussion')}
          </Alert>
        )}
      </>
    );
  }
);

export default MentionTextArea;
