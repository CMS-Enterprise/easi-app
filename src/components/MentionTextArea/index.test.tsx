import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import MentionTextArea from '.';

const content =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.';

describe('MentionTextArea component', () => {
  it('truncates text with read more/less button', async () => {
    render(
      <MentionTextArea
        id="mentionTextArea"
        initialContent={content}
        truncateText
      />
    );

    const truncatedText = content.slice(0, 275);

    expect(screen.getByText(`${truncatedText} ...`)).toBeInTheDocument();

    const readMoreButton = screen.getByRole('button', { name: 'Read more' });
    expect(readMoreButton).toBeInTheDocument();
    userEvent.click(readMoreButton);

    const readLessButton = await screen.findByRole('button', {
      name: 'Read less'
    });
    expect(readLessButton).toBeInTheDocument();

    expect(screen.getByText(content)).toBeInTheDocument();
  });
});
