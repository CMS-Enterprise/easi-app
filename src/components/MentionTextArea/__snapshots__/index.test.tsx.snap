// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MentionTextArea component > renders the component to view text 1`] = `
<DocumentFragment>
  <div
    class="tiptap__readonly font-body-sm line-height-body-5"
  >
    <div
      aria-label="Rich-Text Editor"
      class="tiptap ProseMirror"
      contenteditable="false"
      id="mentionTextArea"
      role="textbox"
      translate="no"
    >
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MentionTextArea component > renders the editable text area component 1`] = `
<DocumentFragment>
  <div
    class="tiptap__editable usa-textarea"
    tabindex="-1"
  >
    <div
      aria-label="Rich-Text Editor"
      class="tiptap ProseMirror ProseMirror-focused"
      contenteditable="true"
      id="mentionTextArea"
      role="textbox"
      tabindex="0"
      translate="no"
    >
      This is the text!
    </div>
  </div>
</DocumentFragment>
`;
